### PR TITLE
Snapshots

### PR DESCRIPTION
**Pull request by birkland**
Adds `make snapshot-image` target to create a snapshot
image, and makes snapshot generation a part of `bootstrap`

Right now, it uses my own personal repository for hosting snapshots, so you can build one via
`make snapshot-image`, but not push it, unless you're me

IDC.md contains documentation on snapshots now.  

One usability addition unrelated to snapshots is a new target `make up`.  This does everything `make` does, plus does a `composer install`.

### CODE REVIEW

File: docker-compose.idc-snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+services:
```
Comment (emetsger) [inline docker-compose.idc-snapshot.yml:2]: needs: `version: "3.7"` at the top

I think all the versions of all the docker compose files need to align (or at least be compatible with each other)

File: snapshot/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,2 @@
+FROM alpine:latest
+COPY data/ /
```
Comment (emetsger) [inline snapshot/Dockerfile:2]: This fails because there's no `data/` directory.  Maybe add and commit `data/.keep`?

File: idc.Makefile [hunk 3]
```diff
@@ -23,3 +29,14 @@ destroy-state:
 composer-install:
 	echo "Installing via composer"
 	docker-compose exec drupal with-contenv bash -lc 'COMPOSER_MEMORY_LIMIT=-1 composer install'
+
+.PHONY: snapshot-image
+.SILENT: snapshot-image
+snapshot-image:
+	docker-compose stop
+	docker run --rm --volumes-from snapshot \
```
Comment (emetsger) [inline idc.Makefile:39]: After a `mkdir snapshot/data` and `docker-compose build snapshot`, this fails:
```
$ make -B snapshot-image
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
Digest: sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321
Status: Image is up to date for alpine:latest
docker: Error response from daemon: No such container: snapshot.
See 'docker run --help'.
make: *** [snapshot-image] Error 125
```

I think the main reason is that no _container_ named `snapshot` exists.  `docker run` is ignorant of `docker-compose.yml`, but `docker-compose run` doesn't support `--volumes-from`.

In order to create one, I had to manually run:
`docker-compose run --name snapshot snapshot`

And then re-invoke `make -B snapshot-image`

**Review by emetsger — CHANGES_REQUESTED** [review]
After creating the missing `data/` directory, and working around and then successfully invoking `make -B snapshot-image`, the snapshot did not appear to take (there's nothing in the database).  So maybe I'm doing something wrong, or maybe there's something more fundamental going on in the PR that I'm not seeing.

**Comment by emetsger** [comment]
So after making the snapshot, I attempt a bootstrap, and I guess this is still in progress:
```
make -B bootstrap
docker-compose pull
Pulling activemq   ... done
Pulling alpaca     ... done
Pulling cantaloupe ... done
Pulling fits       ... done
Pulling crayfits   ... done
Pulling gemini     ... done
Pulling homarus    ... done
Pulling houdini    ... done
Pulling hypercube  ... done
Pulling milliner   ... done
Pulling recast     ... done
Pulling snapshot   ... error
Pulling mariadb    ... done
Pulling matomo     ... done
Pulling drupal     ... done
Pulling solr       ... done
Pulling traefik    ... done

ERROR: for snapshot  pull access denied for idc/snapshot, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
ERROR: pull access denied for idc/snapshot, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
make: *** [pull] Error 1
```

File: docker-compose.idc-snapshot.yml (hunk 1 above)
Comment (birkland) [reply docker-compose.idc-snapshot.yml:2]: Fixed.  By the way, what is the observable effect of this?  It seems to make no difference for me, unless I'm missing something

File: snapshot/Dockerfile (hunk 2 above)
Comment (birkland) [reply snapshot/Dockerfile:2]: there aren't several subdirectories?
```
birkland@MSEL-DRCC07:~/projects/test/idc-dev/local/snapshot/data$ ls
drupal  mariadb-data  mariadb-files  matomo-config  minio  solr
```

What do you see, and what is the error?

Comment (birkland) [reply snapshot/Dockerfile:2]: oh, I must not have comitted them!

File: idc.Makefile (hunk 3 above)
Comment (birkland) [reply idc.Makefile:39]: Fascinating.  There is a docker compose directive:
```
  snapshot:
    build:
      context: /home/birkland/projects/idc-isle-dc/snapshot
    container_name: snapshot
```

Does it ignore that and name the snapshot container something else?

**Comment by birkland** [comment]
Try now.  I think it was to much WIP.   Lots of changes and a README added, so I think it now reflects a more complete picture of what snapshots are supposed to do.

File: idc.Makefile (hunk 3 above)
Comment (emetsger) [reply idc.Makefile:39]: Your command is invoking `docker`, not `docker-compose`, so nothing in `docker-compose.yml` is used.

Comment (birkland) [reply idc.Makefile:39]: That's correct, it's invoking docker to launch a container that dumps a tar file.  However, as an argument (`--volumes-from`), it is referencing the name of a container that is presumed to have been created by docker-compose.  docker-compose has been instructed (via `container_name`) to create a container with that name.  So the question is:  why isn't a container with that name available to you?

my theory is that it's one of the following:
* the stack wasn't running
* docker-compose is creating the `snapshot` container under a different name than is specified via `container_name`.

I haven't seen this, so I'm curious what is happening for you

File: docker-compose.idc-snapshot.yml (hunk 1 above)
Comment (emetsger) [reply docker-compose.idc-snapshot.yml:2]: Building the snapshot image failed because the `COPY data` instruction fails

**Comment by emetsger** [comment]
Updates to iDC-specific Make targets, including the support of snapshot image creation.

**Pull request merged**